pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java

```diff
@@ -743,6 +747,97 @@ public synchronized LedgerOffloader createManagedLedgerOffloader(ServiceConfigur
         }
     }
 
+    public synchronized LedgerOffloader createManagedLedgerOffloader(ServiceConfiguration conf, OffloadPolicies offloadPolicies)
+            throws PulsarServerException {
+        try {
+            if (StringUtils.isNotBlank(offloadPolicies.getDriver())) {
+                checkNotNull(conf.getOffloadersDirectory(),
+                        "Offloader driver is configured to be '%s' but no offloaders directory is configured.",
+                        conf.getManagedLedgerOffloadDriver());
+                this.offloaderManager = OffloaderUtils.searchForOffloaders(conf.getOffloadersDirectory());
+                LedgerOffloaderFactory offloaderFactory = this.offloaderManager.getOffloaderFactory(offloadPolicies.getDriver());
+                try {
+                    return offloaderFactory.create(
+                            conf.getProperties(),
+                            offloadPolicies,
+                            ImmutableMap.of(
+                                    LedgerOffloader.METADATA_SOFTWARE_VERSION_KEY.toLowerCase(), PulsarVersion.getVersion(),
+                                    LedgerOffloader.METADATA_SOFTWARE_GITSHA_KEY.toLowerCase(), PulsarVersion.getGitSha()
+                            ),
+                            getOffloaderScheduler(conf));
+                } catch (IOException ioe) {
+                    throw new PulsarServerException(ioe.getMessage(), ioe.getCause());
+                }
+            } else {
+                LOG.info("No ledger offloader configured, using NULL instance");
+                return NullLedgerOffloader.INSTANCE;
+            }
+        } catch (Throwable t) {
+            throw new PulsarServerException(t);
+        }
+    }
+
+    public Map<NamespaceName, LedgerOffloader> createNamespacesOffloaders() {
```

Comment:
I don't think we should create all the offloaders at the same time. We should create the offloader on-demand for two reasons:

1) save the memory usage. because a broker doesn't own all the namespaces.
2) the namespace policy can be updated at any time. the offloader should reflect to the policies dynamically.

Reply:
Good point. Where is running the offload stuff and the conditions to run it?

Reply:
I think I did it in 2e227e5.